dora/core/server/worker/src/main/java/alluxio/worker/s3/S3NettyHandler.java

```diff
@@ -0,0 +1,698 @@
+import java.net.URLDecoder;
+import java.nio.channels.FileChannel;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+import javax.security.auth.Subject;
+
+/**
+ * S3 Netty Handler - handle http requests from S3 API in Netty.
+ */
+public class S3NettyHandler {
+  private static final Logger LOG = LoggerFactory.getLogger(S3NettyHandler.class);
+  private String mUser;
+  private final String mBucket;
+  private final String mObject;
+  private final FullHttpRequest mRequest;
+  private HttpResponse mResponse;
+  private final ChannelHandlerContext mContext;
+  private final QueryStringDecoder mQueryDecoder;
+  private S3NettyBaseTask mS3Task;
+  private FileSystem mFsClient;
+  private DoraWorker mDoraWorker;
+  private Stopwatch mStopwatch;
+  public AsyncUserAccessAuditLogWriter mAsyncAuditLogWriter;
+  private final FileTransferType mFileTransferType;
+
+  public static final Pattern BUCKET_PATH_PATTERN = Pattern.compile("^" + "/[^/]*$");
+  public static final Pattern OBJECT_PATH_PATTERN = Pattern.compile("^" + "/[^/]*/.*$");
+  public static final boolean BUCKET_NAMING_RESTRICTION_ENABLED =
+      Configuration.getBoolean(PropertyKey.PROXY_S3_BUCKET_NAMING_RESTRICTIONS_ENABLED);
+  public static final WritePType S3_WRITE_TYPE =
+      Configuration.getEnum(PropertyKey.PROXY_S3_WRITE_TYPE, WriteType.class).toProto();
+  // https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucketnamingrules.html
+  // - Undocumented edge-case, no adjacent periods with hyphens, i.e: '.-' or '-.'
+  public static final Pattern BUCKET_ADJACENT_DOTS_DASHES_PATTERN = Pattern.compile("([-\\.]{2})");
+  public static final Pattern BUCKET_INVALIDATION_PREFIX_PATTERN = Pattern.compile("^xn--.*");
+  public static final Pattern BUCKET_INVALID_SUFFIX_PATTERN = Pattern.compile(".*-s3alias$");
+  public static final Pattern BUCKET_VALID_NAME_PATTERN =
+      Pattern.compile("[a-z0-9][a-z0-9\\.-]{1,61}[a-z0-9]");
+  public static final int BUCKET_PATH_CACHE_SIZE = 65536;
+  /* BUCKET_PATH_CACHE caches bucket path during specific period.
+   BUCKET_PATH_CACHE.put(bucketPath,true) means bucket path exists.
+   BUCKET_PATH_CACHE.put(bucketPath,false) plays the same effect
+   as BUCKET_PATH_CACHE.remove(bucketPath). */
+  public static final Cache<String, Boolean> BUCKET_PATH_CACHE = CacheBuilder.newBuilder()
+      .maximumSize(BUCKET_PATH_CACHE_SIZE)
+      .expireAfterWrite(
+          Configuration.global().getMs(PropertyKey.PROXY_S3_BUCKETPATHCACHE_TIMEOUT_MS),
+          TimeUnit.MILLISECONDS)
+      .build();
+  private static final int PACKET_LENGTH = 8 * 1024;
+  private static final String[] UNSUPPORTED_SUB_RESOURCES = {"acl", "policy", "versioning", "cors",
```

Comment:
@Jackson-Wang-7 Thanks for this worker! It make s3 api easy to used, but I wonder what will happen if some of these properties within request?